### PR TITLE
Port attributes for MDIX (Medium Dependent Interface Crossover), Media config and for Latency

### PR DESCRIPTION
**Pull request by tarakreddy1984**

### CODE REVIEW

**Review by tarakreddy1984 — COMMENTED** [review]
build issue

**Review by tarakreddy1984 — COMMENTED** [review]
build fix

**Comment by kcudnik** [comment]
please fix all the build errors, and add empty line before @type to make it distinct from @brief

**Review by tarakreddy1984 — COMMENTED** [review]
build issue

**Review by tarakreddy1984 — COMMENTED** [review]
build issue

**Review by tarakreddy1984 — COMMENTED** [review]
build issue

File: inc/saiswitch.h [hunk 1]
```diff
@@ -2530,6 +2530,30 @@ typedef enum _sai_switch_attr_t
      * @flags READ_ONLY
      */
     SAI_SWITCH_ATTR_AVAILABLE_DOUBLE_NAPT_ENTRY,
+    
+    /**
+     * @brief Parallel processing attributes to lock shared resources to achieve thread/process synchronization
+     *
+     *  The attributes data type is void pointer and the format of the function pointers are as follows:
+     *  int brcm_mutex_lock(uint64_t platform_context)
```
Comment (kcudnik) [inline inc/saiswitch.h:2538]: are those callbacks ?, why they are broadcom specific ?
if so they would need to be specifically added as SAI_SWITCH_ATTR_REGISTER_WRITE, and parse.pl needs to be updated

Comment (tarakreddy1984) [reply inc/saiswitch.h:2538]: I will take out these two attributes and will add as separate PR

**Review by tarakreddy1984 — COMMENTED** [review]
Removed saiswitch changes,

**Review by tarakreddy1984 — COMMENTED** [review]
Removed saiswitch changes,

**Review by tarakreddy1984 — COMMENTED** [review]
build issue

**Review by tarakreddy1984 — COMMENTED** [review]
build issue

**Comment by kcudnik** [comment]
please check build warnings and errors https://dev.azure.com/mssonic/build/_build/results?buildId=9795&view=logs&j=83516c17-6666-5250-abde-63983ce72a49&t=00be4b52-4a63-5865-8e02-c61723ad0692

**Review by tarakreddy1984 — COMMENTED** [review]
build fix

**Review by tarakreddy1984 — COMMENTED** [review]
Build fix

**Review by tarakreddy1984 — COMMENTED** [review]
Build fix

**Review by tarakreddy1984 — COMMENTED** [review]
build fix

**Review by tarakreddy1984 — COMMENTED** [review]
build fix

**Review by kcudnik — APPROVED** [review]

**Comment by jleveque** [comment]
@tarakreddy1984: Can you please add a description to this PR?

File: inc/saiport.h [hunk 2]
```diff
@@ -1698,6 +1765,95 @@ typedef enum _sai_port_attr_t
...
+    SAI_PORT_ATTR_AUTO_MEDIUM,
+
+    /**
+     * @brief Configure media precedence Fiber preferred -> True, Copper preferred -> False
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default false
+     */
+    SAI_PORT_ATTR_MEDIA_PREFERRED,
```
Comment (SathishKumarTR-Dell) [inline inc/saiport.h:1837]: Please capture what is the expected sequence to be followed with these attribute if an application requires the medium "auto-detect" to be disabled and initialize the port in "fiber" mode by default.

File: inc/saiport.h [hunk 3]
```diff
@@ -1698,6 +1765,95 @@ typedef enum _sai_port_attr_t
...
+    SAI_PORT_ATTR_POWER_MODE,
+
+    /**
+     * @brief Enabled auto detection between 1000X and SGMII slave mode
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default false
+     */
+    SAI_PORT_ATTR_AUTO_DETECT,
```
Comment (SathishKumarTR-Dell) [inline inc/saiport.h:1810]: It looks like there is some dependency with the AUTO_MEDIUM attribute proposed below. For e.g, is this configurable when AUTO_MEDIUM is set to false?

File: inc/saiport.h [hunk 4]
```diff
@@ -1698,6 +1765,95 @@ typedef enum _sai_port_attr_t
...
+    SAI_PORT_ATTR_MDIX_MODE,
+
+    /**
+     * @brief Configure Auto Neg mode
+     *
+     * @type sai_port_auto_neg_config_mode_t
+     * @flags CREATE_AND_SET
+     * @default SAI_PORT_AUTO_NEG_CONFIG_MODE_DISABLED
+     */
+    SAI_PORT_ATTR_AUTO_NEG_CONFIG_MODE,
```
Comment (SathishKumarTR-Dell) [inline inc/saiport.h:1792]: Is this applicable only for the 1000BASE-T ports? Can you please call out what port speeds this attribute is actually relevant?

**Comment by rlhui** [comment]
Could we please add short description of this proposal?  Thanks.

**Comment by rlhui** [comment]
@jleveque  - would you please help review, thanks?

File: inc/saiport.h (hunk 4 above)
Comment (rlhui) [reply inc/saiport.h:1792]: @SathishKumarTR-Dell , thanks, please help to review/sign-off this PR?

**Review by jleveque — APPROVED** [review]

**Review by tarakreddy1984 — COMMENTED** [review]
updated review comments

**Review by tarakreddy1984 — COMMENTED** [review]
doc error fix

File: inc/saiport.h [hunk 5]
```diff
@@ -1698,6 +1749,82 @@ typedef enum _sai_port_attr_t
...
+
+    /**
+     * @brief Enabled auto detection between 1000X and SGMII slave mode
```
Comment (apogedik) [inline inc/saiport.h:1779]: Enabled -> Enable

File: inc/saiport.h [hunk 6]
```diff
@@ -1698,6 +1749,82 @@ typedef enum _sai_port_attr_t
...
+     * @default false
+     */
+    SAI_PORT_ATTR_AUTO_MEDIUM,
+
+    /**
+     * @brief Configure media precedence.
+     *
+     * The PHY supporting mutually exclusive copper and fiber ports should be
+     * configured to select the preferred/active media.
+     * True will configure Fiber ports to be active and False will configure
```
Comment (apogedik) [inline inc/saiport.h:1810]: * True:  Configure Fiber ports to be active 
* False Configure Copper ports to be active

File: meta/acronyms.txt [hunk 7]
```diff
@@ -146,3 +146,5 @@ PRBS - Pseudorandom binary sequence
 NAPT - Network Address Port Translation
 SNAPT - Source Network Address Port Translation
 DNAPT - Destination Network Address Port Trasnlation
+SGMII - Serial gigabit media-independent interface
```
Comment (apogedik) [inline meta/acronyms.txt:149]: SGMII - Serial Gigabit Media Independent Interface

**Review by tarakreddy1984 — COMMENTED** [review]
Updated doc section

**Comment by rlhui** [comment]
> @tarakreddy1984: Can you please add a description to this PR?

@tarakreddy1984  - description PR needed, thanks.

**Comment by tarakreddy1984** [comment]
Facing some merge issue and would not able revert the old changes.  Will create new PR.
Sorry for inconvenience

**Review by tarakreddy1984 — COMMENTED** [review]
Updated with fresh commit

**Review by tarakreddy1984 — COMMENTED** [review]
Updated files

**Review by tarakreddy1984 — COMMENTED** [review]
spell fix

**Review by tarakreddy1984 — COMMENTED** [review]
spell fix

**Review by tarakreddy1984 — COMMENTED** [review]
spell fix

File: inc/saiport.h (hunk 3 above)
Comment (SidharajU) [reply inc/saiport.h]: There is no relation/dependency between auto medium and auto detect. Auto-detect allows auto detection between SGMIi-Slave and 1000BaseX modules and Auto-medium allows both fiber and copper ports to be active at the same time. If auto-medium is set to false, then only copper medium is active and the value of auto-detect is ignored. Auto-detect value is only valid when SAI_PORT_ATTR_MEDIA_PREFERRED is set to true to allow Fiber medium to take precedence

File: inc/saiport.h (hunk 2 above)
Comment (SidharajU) [reply inc/saiport.h]: By default auto-detect will be disabled. If its enabled during prior execution, then SAI_PORT_ATTR_AUTO_DETECT attribute can be set to false to disable it. To initialize fiber ports, SAI_PORT_ATTR_MEDIA_PREFERRED should be set to true to use Fiber interface

**Review by tarakreddy1984 — COMMENTED** [review]
updated with module type

**Review by tarakreddy1984 — COMMENTED** [review]
Updated spell fix

File: inc/saiport.h [hunk 8]
```diff
@@ -1698,6 +1765,85 @@ typedef enum _sai_port_attr_t
...
+     * @brief Configure EEE latency mode
+     *
+     * False: Fixed latency
+     * True: Variable latency
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default false
+     */
+    SAI_PORT_ATTR_EEE_LATENCY_MODE,
```
Comment (rlhui) [inline inc/saiport.h:1836]: there is another SAI_PORT_ATTR_LOW_LATENCY_VARIATION, being introduced in https://github.com/opencomputeproject/SAI/pull/1209/ are they related or independent?

Comment (SidharajU) [reply inc/saiport.h:1836]: No. That's different. This is used in Energy Efficient Ethernet feature and only applicable to Copper PHY. Other attribute is applicable to high speed optical PHYs.

File: inc/saiport.h (hunk 3 above)
Comment (rlhui) [reply inc/saiport.h]: @SathishKumarTR-Dell  - please review/approve.  There is also a doc included in this PR.

File: inc/saiport.h (hunk 4 above)
Comment (lguohan) [reply inc/saiport.h:1792]: SAI_PORT_ATTR_AUTO_NEG_MODE

File: inc/saiport.h [hunk 9]
```diff
@@ -1698,6 +1765,85 @@ typedef enum _sai_port_attr_t
...
+
+    /**
+     * @brief Configure Fiber module type
+     *
+     * @type sai_port_module_type_t
+     * @flags CREATE_AND_SET
+     * @default SAI_PORT_MODULE_TYPE_1000BASE_X
+     */
+    SAI_PORT_ATTR_MODULE_TYPE,
```
Comment (lguohan) [inline inc/saiport.h:1845]: how is it different from SAI_PORT_ATTR_INTERFACE_TYPE

**Comment by tarakreddy1984** [comment]
withdrawn this PR

**Pull request closed without merging**